vowpalwabbit/cats_tree.cc

```diff
@@ -0,0 +1,385 @@
+          float trained_action = (ec.pred.scalar < 0) ? LEFT : RIGHT;
+          if (trained_action == local_action)
+          {
+            cost_parent =
+              (std::min)(cost_v, cost_w) * fabs(ec.pred.scalar) + (std::max)(cost_v, cost_w) * (1 - fabs(ec.pred.scalar));
+            VW_DBG(ec) << "otree_c: learn() ec.pred.scalar == local_action" << std::endl;
+          }
+          else
+          {
+            cost_parent =
+              (std::max)(cost_v, cost_w) * fabs(ec.pred.scalar) + (std::min)(cost_v, cost_w) * (1 - fabs(ec.pred.scalar));
+            VW_DBG(ec) << "otree_c: learn() ec.pred.scalar != local_action" << std::endl;
+          }
+        }
+
+      }
+      if (i == 0)
+        a_parent_cost = cost_parent;
+      else
+        b_parent_cost = cost_parent;
+    }
+    _a = {nodes[_a.node_id].parent_id, a_parent_cost};
+    _b = {nodes[_b.node_id].parent_id, b_parent_cost};
+  }
+
+  ec.l = saved_label;
+  ec.weight = saved_weight;
+  ec.pred = saved_pred;
+}
+
+void cats_tree::set_trace_message(std::ostream* vw_ostream) { _trace_stream = vw_ostream; }
+
+cats_tree::~cats_tree()
+{
+  if(_trace_stream != nullptr)
+    (*_trace_stream) << tree_stats_to_string() << std::endl;
+}
+
+std::string cats_tree::tree_stats_to_string()
+{
+  return _binary_tree.tree_stats_to_string();
+}
+
+void predict(cats_tree& ot, single_learner& base, example& ec)
+{
+  VW_DBG(ec) << "otree_c: before tree.predict() " << multiclass_pred_to_string(ec) << features_to_string(ec)
+             << std::endl;
+  ec.pred.multiclass = ot.predict(base, ec);
+  VW_DBG(ec) << "otree_c: after tree.predict() " << multiclass_pred_to_string(ec) << features_to_string(ec)
+             << std::endl;
+}
+
+void learn(cats_tree& tree, single_learner& base, example& ec)
+{
+  VW_DBG(ec) << "otree_c: before tree.learn() " << cb_label_to_string(ec) << features_to_string(ec) << std::endl;
+  tree.learn(base, ec);
+  VW_DBG(ec) << "otree_c: after tree.learn() " << cb_label_to_string(ec) << features_to_string(ec) << std::endl;
+}
+
+base_learner* setup(options_i& options, vw& all)
+{
+  option_group_definition new_options("CATS Tree Options");
+  uint32_t num_actions; // = K = 2^D
+  uint32_t bandwidth; // = 2^h#
+  uint32_t scorer_flag;
+  new_options.add(make_option("cats_tree", num_actions).keep().help("CATS Tree with <k> labels"))
+      .add(make_option("scorer_option", scorer_flag)
+               .default_value(0)
+               .keep()
+               .help("CATS Tree reduction to scorer [-1, 1] versus binary -1/+1"))
+      .add(make_option("bandwidth", bandwidth)
+               .default_value(0)
+               .keep()
+               .help("bandwidth for continuous actions in terms of #actions"));
+
+  options.add_and_parse(new_options);
+
+  if (!options.was_supplied("cats_tree"))
+    return nullptr;
+
+  if (scorer_flag)
+  {
+    options.insert("link", "glf1");
+  }
+  else  // if (!options.was_supplied("binary"))
```

Comment:
Is this comment relevant? Either remove or fix

Reply:
Removed.